### DISCUSSION TITLE
AOT compilation of machines

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2456] [Ideas]
Hi. This might be early to discuss while you guys are busy with v5. I just would like to add it to the wish list.

The goal is to improve performance by shifting machine interpretation to build time and minimize bundle size by eliminating larger machine definitions.

This is especially relevant when we would like to build entire/larger applications based on xstate machines.

**Comment by ChrisShank** [comment]
Just curious are you hit a performance wall with xstate? If so where? Moving `xstate` from runtime to compile-time is likely to be quite a change and I am personally not sure that it's inherently a good thing to do.

In the context of web apps, IME state machines *typically* dont have the largest performance overhead (web based games are a different topic though). It’s quite a nuanced discussion though because “good performance” can mean multiple things and is typically a balancing act. So i think we first have to define what kinds of performance we are talking about here and how to
measure it!

For example in web apps one way to view performance is by how much your app blocks the main thread. Blocking it causes user interactions to be delayed, animation to be laggy, ec.t All of these factors contribute to a *perpetually* slow application. If you are talking about performance in this context, then AOT performance is not the *only* way to improve performance and for the tradeoffs might not be the best! I think there are a couple of improvements on this front:

1. Improve raw JS execution. 
  - One way to improve runtime execution is write a more efficient JS. This is very [reminiscent](https://twitter.com/_developit/status/1412451442946981890) of what Jason Miller's is exploring with `preact`, basically how efficient can we write JS based on how it is compiled. The thing with this is that you *will* hit a wall with this approach.
  - Alternatively you can create tooling to AOT compile machines/interpreters and take advantage of possible build time analysis to create really efficient and potentially small machines. This is very similar to `svelte`'s approach where you effectively compile away the runtime to generate small and performant bundles. Now the interesting thing about compiling away the runtime creates an inflection point where lots of components `svelte` (or as you propose machines in `xstate`) creates a larger bundle size that having a runtime. Evan You [explores](https://twitter.com/youyuxi/status/1411146794747191296) this pretty deeply when comparing `svelte` vs. `vue`.

2. Add more scheduling for actors
  - This is *sort of* similar approach to what `react` concurrent mode is exploring, cooperative single-threaded scheduling. This approach is trying to do work in small chunks that yields back to the browser. The cool thing is that `xstate` can take advantage of the actor model that it was built on. All actors have a queue of messages and they read those messages one at a time. In the future, xstate could provide more granular scheduling of when messages are processed by actors so that one actor does not block the main thread by reading all of its messages at once. My understanding (which is probably wrong) is that actors process new messages immediately ([see simple example](https://github.com/statelyai/xstate/blob/main/packages/core/src/behaviors.ts#L127)), which could block the main thread.

3. Move behavior to web workers
  - What if could move behavior off the main thread entirely? With web workers we can do that, and `xstate` can easily have first-class support for web workers if we use the actor model. I have a POC of this [here](https://github.com/ChrisShank/xstate-isomorphic-behavior/blob/master/src/main.ts#L8). The general idea is that you can directly invoke or spawn a web worker like `invoke: { src: invokeWebWorker('./worker.js') }` and send and receive messages from it. this approach lets you seamlessly take a machine that was running on the main thread and trivially move it to a worker. Out biggest limitation now is the time it takes to copy data between threads, but in most cases [there is not much overhead](https://surma.dev/things/is-postmessage-slow/index.html).

Just some of my musing on the subject, what are your thoughts?

**Comment by Andarist** [comment]
AOT is something to be explored on our roadmap - but, quite frankly, it doesn't have the highest priority. It's just something nice to have but considering the amount of work needed for this to be done it won't happen anytime soon.

My main motivation for wanting to have this someday is not as much about the performance as it is about bundlesize. Currently, it's hard to justify using XState (as a runtime) for reusable libraries that integrate with any app. If the library is 3kb then adding an additional 10kb just to use XState isn't a justifiable choice.
